### ISSUE TITLE
Determine when an AGV leaves a node

### ISSUE THREAD

**Issue by HardyBenjamin** [VDA5050/VDA5050#273]
Hello,

How determine when an AGV leaves a node ?
When an AGV arrives in the deviation range of a node, it removes the node from the nodeState array and it removes the edge it was on from the edgeState array.
So how do you determine when an AGV leaves the devition range of a node?

Thanks,

**Comment by AntonDueck** [comment]
only if it enters the deviation of the next node.  because then the next state is at least generated. you could use the visualization data and do the math yourself when the AGV leaves the node, or if there is another trigger to send the state und take the position from the state.

**Comment by HardyBenjamin** [comment]
As shown in the figure: "Figure 13 nodeStates, edgeStates, actionStates during order handling", When an AGV arrives in the deviation range of a node, it removes the node from the nodeState array and it removes the edge it was on from the edgeState array.
So the arrays nodeStates and EdgeStates give the same information. Wouldn't it be more interesting if the 2 arrays worked from the same function to get additional information? When an AGV arrives in the deviationRange of a node, it removes the node from the node array. And when it arrives at an edge (leaves the divationRange of a node), it removes the edge from the edge array.

**Comment by collani-bosch** [comment]
Be aware of chapter 6.10.2: "The AGV decides on its own when a node should count as traversed. Generally, the AGV’s control point should be within the node’s deviationRangeXY and its orientation within deviationRangeTheta ."
Entering the deviation range is a necessary requirement to count the node as traversed not a sufficient requirement. One vehicle count the node as traversed on entering the deviation range another vehicle has to meet additional requirements. Or the other way around: entering and / or leaving the deviation range of a node alone isn't a trigger to send a new state message.
Some thoughts: 
- I don't know your use-case but you seam to be interested at two poses, not only in one. Maybe operating with two nodes fits better.
- For the communication between the vehicle and MC you have no timing guarantees. The delay and jitter maybe are larger than expected and required.
- The state message isn't really an event. The state message shall describe the state of the vehicle completely, because you can loose messages (MQTT QoS level 0).
- What would be the behavior, if the vehicle enters the deviation range of two different nodes? Now things became really complicated.

**Comment by camil22** [comment]
I agree with @HardyBenjamin  that currently there's not much data in the state indicating that the agv started to drive on the next edge vs still being on the node. If I'm not mistaken distanceSinceLastNode or agvPosition is the only data indicating that the agv has started to drive on the next edge, but both are not mandatory.
If MC wants to act upon an agv leaving a node (accepting state message loss of course) the only way to safely communicate it is on the next node when the edge gets removed from the `edgeStates` array.
I'd also like to see a distinction between reaching a node and entering the next edge in the state message.

**Comment by Mehno** [comment]
If we change the `lastNodeSequenzID` to a `lastSequenzID` we could resolve cases like this.

**Comment by ConstantinUp** [comment]
Please check, if with version 3.0 the expected behavior is more clear. 
If the robot should execute something when entering the next edge instead of on the node, an edge action with an appropriate blocking type would be a possibility.

In addition to the proposal by @collani-bosch, can you please provide a concrete example and proposal what to change best in form of a PR.
